arches/app/src/components/ControlledListManager/ImageEditor.vue

```diff
@@ -0,0 +1,231 @@
+<script setup lang="ts">
+import arches from "arches";
+import { computed, inject, ref } from "vue";
+import { useGettext } from "vue3-gettext";
+
+import Button from "primevue/button";
+import Column from "primevue/column";
+import Dropdown from "primevue/dropdown";
+import DataTable from "primevue/datatable";
+import InputText from "primevue/inputtext";
+import { useToast } from "primevue/usetoast";
+
+import AddMetadata from "@/components/ControlledListManager/AddMetadata.vue";
+
+import { itemKey } from "@/components/ControlledListManager/const.ts";
+import { deleteImage, upsertMetadata, deleteMetadata } from "@/components/ControlledListManager/api.ts";
+import { bestLabel, languageName } from "@/components/ControlledListManager/utils.ts";
+
+import type { Ref } from "vue";
+import type { DataTableRowEditInitEvent } from "primevue/datatable";
+import type {
+    ControlledListItem,
+    ControlledListItemImage,
+    ControlledListItemImageMetadata,
+    NewControlledListItemImageMetadata,
+} from "@/types/ControlledListManager";
+
+const { image, removeImage, appendImageMetadata, updateImageMetadata, removeImageMetadata } = defineProps<{
+    image: ControlledListItemImage;
+    removeImage: (removedImage: ControlledListItemImage) => void;
+    appendImageMetadata: (appendedMetadata: ControlledListItemImageMetadata | NewControlledListItemImageMetadata) => void;
+    updateImageMetadata: (updatedMetadata: ControlledListItemImageMetadata) => void;
+    removeImageMetadata: (removedMetadata: ControlledListItemImageMetadata | NewControlledListItemImageMetadata) => void;
+}>();
+const item = inject(itemKey) as Ref<ControlledListItem>;
+
+const editingRows = ref([]);
+
+const DANGER = "danger";
+const toast = useToast();
+const { $gettext } = useGettext();
+
+const METADATA_CHOICES = [
+    {
+        type: 'title',
```

Comment:
Hey, would you mind fleshing this out for me a bit more? This needs to be object-oriented to drive a primevue dropdown and also support dynamic lookups. But I think I might be missing the gist of your thought.

Reply:
Absolutely, similar to L39 here. Essentially trying to avoid magic strings ( or really _any_ context-less strings ) being used in the code. Admittedly this may be taking a bit of a hard-line stance